// wdio-lambdatest-service/build/util.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getParentSuiteName = getParentSuiteName;
exports.updateSessionById = updateSessionById;
require("source-map-support/register");
const axios = require('axios');
const {
  version,
  appVersion,
  baseUrl,
  baseUrlApp
} = require("./constants.js");
const logger = require('./logger')(process.env.LT_API_LOG_FILE);
function getParentSuiteName(fullTitle, testSuiteTitle) {
  const fullTitleWords = fullTitle.split(' ');
  const testSuiteTitleWords = testSuiteTitle.split(' ');
  const shortestLength = Math.min(fullTitleWords.length, testSuiteTitleWords.length);
  let c = 0;
  let parentSuiteName = '';
  while (c < shortestLength && fullTitleWords[c] === testSuiteTitleWords[c]) {
    parentSuiteName += fullTitleWords[c++] + ' ';
  }
  return parentSuiteName.trim();
}
async function updateSessionById(sessionId, data, lambdaCredentials) {
  const sessionUrl = lambdaCredentials.isApp ? `${baseUrlApp}${appVersion.latestVersion}/sessions/${sessionId}` : `${baseUrl}${version.latestVersion}/sessions/${sessionId}`;
  let config = {
    method: 'patch',
    maxBodyLength: Infinity,
    url: sessionUrl,
    headers: {
      'accept': 'application/json',
      'Authorization': `Basic ${Buffer.from(lambdaCredentials.username + ':' + lambdaCredentials.accessKey).toString('base64')}`,
      'Content-Type': 'application/json'
    },
    data: data
  };
  try {
    var _response$config;
    let response = await axios.request(config);
    logger.info((response === null || response === void 0 || (_response$config = response.config) === null || _response$config === void 0 ? void 0 : _response$config.data) + (response === null || response === void 0 ? void 0 : response.data));
  } catch (error) {
    logger.error(error);
  }
}
